### DISCUSSION TITLE
Add EVMole integration for unverified contracts

### DISCUSSION THREAD

**Discussion by cdump** [BuidlGuidl/abi.ninja#102] [Ideas]
Currently, abi.ninja is trying to use heimdall-rs for ABI extraction from unverified contracts.

I propose to add (or replace) it with [EVMole](https://github.com/cdump/evmole), which has the following advantages:

- Better accuracy in [function selectors extraction](https://github.com/cdump/evmole?tab=readme-ov-file#benchmark)
- Significant improvement in accuracy for [function arguments extraction](https://github.com/cdump/evmole?tab=readme-ov-file#function-arguments)
- Faster performance
- A JavaScript implementation with zero dependencies, allowing it to be embedded in abi.ninja without an external backend

It already has an [online demo](https://cdump.github.io/evmole/). You can see an example usage in the source code of that page (right inside the `<script>` tag).

My proposed logic for unverified contracts:

1. Load runtimeBytecode (requires an external RPC call to `eth_getCode`; 1rpc is used in the EVMole demo)
2. Get function selectors with the `evmole.functionSelectors` JS function (offline)
3. Resolve function selectors using an API, like https://openchain.xyz/signatures
4. For unresolved selectors from the previous step, extract arguments using `evmole.functionArguments` (offline)

What do you think about this?

**Comment by cdump** [comment]
@technophile-04 @portdeveloper what do you think about this?

**Reply by portdeveloper** [reply]
Hey I think it looks great but how many chains does EVMole support right now?

**Comment by cdump** [comment]
All EVM chains, because evmole doesn't interact with the chain, only the
evm bytecode required

**Reply by portdeveloper** [reply]
That definitely sounds interesting!

**Comment by portdeveloper** [comment]
Alright just checked how it works and it definitely is really cool!
Here is a random SE2 app i spinned up to test it. https://test-evmole.vercel.app/

I have some questions:
-> Is it possible for evmole to return function types?
-> Is it possible to get an abi-like result from evmole?

**Comment by cdump** [comment]
1. Are you talking about function return type (`outputs` in abi)? It's not extracted by evmole yet.
2. 
```javascript
const info = contractInfo(code, { selectors: false, arguments: true, stateMutability: true });
const abi = info.functions.map((fn) => ({
  type: 'function',
  name: `func_${fn.selector}`,
  stateMutability: fn.stateMutability,
  outputs: [], // or undefined - not extracted by evmole
  inputs: [], // need to parse arguments string with some library or code
}));
```
I can't recommend any particular JS library for converting arguments into inputs, but I can say that this arguments string is constructed using the same rules as those used in function signature hashes. So, `keccak('${functionName}(${evmoleArguments)')[:4]`  represents the function signature
